lib/pbench/server/api/resources/server_configuration.py

```diff
@@ -0,0 +1,152 @@
+                #       {"config": {"dataset-lifetime": "2y"}}
+                #
+                # Some options may not make sense: need to think about this.
+                query_schema=Schema(Parameter("value", ParamType.STRING)),
+                body_schema=Schema(
+                    Parameter("config", ParamType.JSON, keywords=ServerConfig.KEYS),
+                    Parameter("value", ParamType.JSON),
+                ),
+                authorization=API_AUTHORIZATION.ADMIN,
+            ),
+            ApiSchema(
+                API_METHOD.GET,
+                API_OPERATION.READ,
+                uri_schema=Schema(
+                    Parameter("key", ParamType.KEYWORD, keywords=ServerConfig.KEYS)
+                ),
+                authorization=API_AUTHORIZATION.NONE,
+            ),
+            always_enabled=True,
+        )
+
+    def _get(self, params: ApiParams, request: Request) -> Response:
+        """
+        Get the values of server configuration parameters.
+
+        Args:
+            params: API parameters
+            request: The original Request object containing query parameters
+
+        GET /api/v1/server/configuration?key=config1&key=config2
+        """
+
+        key = params.uri.get("key")
+        try:
+            if key:
+                s = ServerConfig.get(key)
+                return jsonify({key: s.value if s else None})
+            else:
+                return jsonify(ServerConfig.get_all())
+        except ServerConfigError as e:
+            raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR, str(e))
+
+    def _put(self, params: ApiParams, _) -> Response:
+        """
+        Set or modify the values of server configuration keys.
+
+        PUT /api/v1/server/configuration
+        {
+            "config": {
+                "dataset-lifetime": 10,
+                "server-state": "running"
+            }
+        }
+
+        PUT /api/v1/server/configuration/dataset-lifetime?value=10
+
+        PUT /api/v1/server/configuration/dataset-lifetime
+        {
+            "value": "10"
+        }
+        """
+
+        try:
+            key = params.uri["key"]
+            try:
+                value = params.query["value"]
+            except KeyError:
+                try:
+                    value = params.body["value"]
+                except KeyError:
+                    raise APIAbort(
+                        HTTPStatus.BAD_REQUEST,
+                        "Incorrect value specification for key {key!r}",
+                    )
+            try:
+                ServerConfig.set(key=key, value=value)
+                return jsonify({key: value})
+            except ServerConfigBadValue as e:
+                raise APIAbort(HTTPStatus.BAD_REQUEST, str(e))
+            except ServerConfigError as e:
+                raise APIAbort(HTTPStatus.INTERNAL_SERVER_ERROR, str(e))
+        except KeyError:
+            pass
+
+        try:
+            config = params.body["config"]
```

Comment:
No, it gets set to `10` because we process `value` first and then `return`.

But if you want to get funky, how about

```
PUT /api/v1/server/configuration/dataset-lifetime?value=5
{
  "config": {
    "dataset-lifetime": 20, "server-state": "running"
  },
  "value": "10"
}
```

So, yeah; I could add code to detect and complain about the redundant extra values, going to extra effort to fail because someone supplied extra junk. That didn't seem worthwhile.

As I'd said, the variations here were somewhat experimental, and feedback on the clarity of the alternatives is definitely reasonable.

Maybe I should drop the `value` query parameter... or drop the `value` JSON body key and require a `PUT` to either specify the full JSON with `/server/configuration` *or* `/server/configuration/key?value=foo`. I was worried about the need to specify JSON values for most of the config options, but I *could* deserialize the query parameter value... it's just a bit ugly since you might need to use HTML quoting to get a JSON object there.

Reply:
I missed the `return` statement.  (I don't know why, but I find `try` blocks to be really hard to read, doubly so when they are nested....)

> That didn't seem worthwhile.

I understand.  But, when I'm on the requesting end, I'm going to want clarity and lack of ambiguity, and good feedback will be much appreciated, especially when I do something wrong.

Personally, I rather like the ability to pass parameters in three ways, potentially in combination.  But, if we enable that, then we need to be clear about what happens if the same parameter is passed in more than one way in the same request.  One approach is to document the order in which the parameters override each other.  But, what I think is better is to insist that the user pass each parameter only once -- that way, the ordering doesn't matter (and we don't have to document it!)...but, it does require more effort in the code.  Still, I think it's worth it.

For simple requests, being able to use the query parameters is very handy; however, for complex requests, that gets cumbersome, hard to read and maintain, etc.  So, I like the idea of being able to pass the value either in a query  parameter or in the body.  Likewise, I like the idea of being able to pass the key either in the URI or, for bulk updates, in the body.  But, I think we should object if the key is passed in both places, because that creates various problems; and, similarly, I think we should object if the value is passed in more than one place.

Let me know if you want me to take a crack at coding that up for you.